core/system/src/unvic.c

```diff
@@ -91,113 +97,86 @@ static int unvic_acl_check(int irqn)
     /* get vector entry */
     uv = &g_unvic_vector[irqn];
 
-    /* an IRQn slot is considered as registered if the ISR entry in the unprivileged
-     * table is not 0 */
-    is_irqn_registered = uv->hdlr ? 1 : 0;
+    if (uv->id == g_active_box) {
+        return UNVIC_ISR_OWNER_SELF;
+    }
+    if (uv->id == UVISOR_BOX_ID_INVALID) {
+        return UNVIC_ISR_OWNER_NONE;
+    }
+    DPRINTF("Disallowed to access IRQ %d from box %d!\n\r", irqn, g_active_box);
+    return UNVIC_ISR_OWNER_OTHER;
+}
 
-    /* check if the same box that registered the IRQn is accessing it
-     * note: if the vector entry for a certain IRQn is 0, it means that no secure
-     *       box claimed exclusive ownership for it. So, another box can claim it
-     *       if it is currently un-registered (that is, if the registered handler
-     *       is NULL) */
-    if(uv->id != g_active_box && (uv->id || is_irqn_registered))
+/* @retval 0 you do not own this interrupt
+ * @retval 1 you own this interrupt
+ */
+static int unvic_isr_register(uint32_t irqn)
+{
+    int isr_status = unvic_acl_check(irqn);
+    switch (isr_status)
     {
-        HALT_ERROR(PERMISSION_DENIED,
-                   "Permission denied: IRQ %d is owned by box %d\n\r", irqn,
-                                                                       uv->id);
+        case UNVIC_ISR_OWNER_NONE:
+            g_unvic_vector[irqn].id = g_active_box;
+            DPRINTF("IRQ %d registered to box %d\n\r", irqn, g_active_box);
+        case UNVIC_ISR_OWNER_SELF:
+            return 1;
+        default:
+            break;
     }
-
-    return is_irqn_registered;
+    return 0;
 }
 
 void unvic_isr_set(uint32_t irqn, uint32_t vector)
 {
     /* verify IRQ access privileges */
-    unvic_acl_check(irqn);
-
-    /* save unprivileged handler
-     * note: if the handler is NULL the IRQn gets de-registered for the current
-     *       box, meaning that other boxes can still use it. In this way boxes
-     *       can share un-exclusive IRQs. A secure box should not do that if it
-     *       wants to hold exclusivity over an IRQn
-     * note: when an IRQ is released (de-registered) the corresponding IRQn is
-     *       disabled, to ensure that no spurious interrupts are served */
-    if (vector) {
-        g_unvic_vector[irqn].id = g_active_box;
-    }
-    else {
-        NVIC_DisableIRQ(irqn);
-        g_unvic_vector[irqn].id = 0;
+    if (!unvic_isr_register(irqn)) {
+        return;
```

Comment:
(If instead you want to keep the semantics by which `unvic_isr_register(irqn)` returns the ownership status of `irqn`, you will have to halt from every function that calls that function. I like this semantics so I'd rather duplicate the `HALT_ERROR`. @Patater ?)